rust/lance-linalg/src/simd/f16.c

```diff
@@ -0,0 +1,53 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stddef.h>
+
+#ifndef LANES
+#define LANES 8
+#endif
+
+/// Works on NEON + FP16 or AVX512FP16
+//
+// Please make sure run "cargo bench --bench norm_l2" on both Apple Silicon and
+// X86_64, before you change this function.
+_Float16 norm_l2_f16(_Float16* data, size_t dimension) {
+  _Float16 sums[LANES];
+#pragma clang loop unroll(enable) vectorize(enable)
+  for (size_t i = 0; i < LANES; i++) {
+    sums[i] = 0;
+  }
+
+#pragma clang loop unroll(enable) vectorize(enable) interleave(enable)
```

Comment:
If we know the SIMD instructions we're shooting for, part of me wonders whether it would be simpler to just write hand-written assembly. There's a nice macro to do this within Rust files: https://doc.rust-lang.org/reference/inline-assembly.html Would let us avoid this whole C thing.

Reply:
Yea, Rob and I briefly discussed this option as well. The generated asm is a bit more complicate to write, because LLVM does forloop unrolling very aggressive, which requires to repeatly write a bunch of same assembly. 

For example, you can check the output via

```
clang -S -mcpu=native <..other-flags> src/simd/f16.c
```